### PR TITLE
microcodeIntel: 20180312 -> 20180807 and iucode-tool: init at 2.3.1

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

Instead of using our vendored version of the microcode processing tool, use iucode-tool. 

I can confirm that at least on one machine, the microcode as processed by this PR is successfully applied.

Cc: @fpletz @wkennington 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/microcode/intel.nix [hunk 1]
```diff
@@ -1,27 +1,26 @@
...
-    gcc -O2 -Wall -o intel-microcode2ucode ${./intel-microcode2ucode.c}
-    ./intel-microcode2ucode microcode.dat
-  '';
-
   installPhase = ''
+    runHook preInstall
+
     mkdir -p $out kernel/x86/microcode
-    mv microcode.bin kernel/x86/microcode/GenuineIntel.bin
+    iucode_tool -w kernel/x86/microcode/GenuineIntel.bin intel-ucode{,-with-caveats}/
```
Comment (Mic92) [inline pkgs/os-specific/linux/microcode/intel.nix]: What is intel-ucode-with-caveats?

File: pkgs/os-specific/linux/microcode/iucode-tool.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  nativeBuildInputs = [ autoreconfHook ];
+
+  enableParallalBuilding = true;
+
+  meta = with stdenv.lib; {
+    description = "Intel® 64 and IA-32 processor microcode tool";
+    homepage = https://gitlab.com/iucode-tool/iucode-tool;
+    license = licenses.gpl2;
+    maintainers = with maintainers; [ peterhoeg ];
+    platforms = platforms.linux;
```
Comment (Mic92) [inline pkgs/os-specific/linux/microcode/iucode-tool.nix]: Can you restrict platforms to x86? The tool uses `cpuid` which is processor specific.

File: pkgs/os-specific/linux/microcode/iucode-tool.nix [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  src = fetchFromGitLab {
+    owner  = "iucode-tool";
+    repo   = "iucode-tool";
+    rev    = "v${version}";
+    sha256 = "04dlisw87dd3q3hhmkqc5dd58cp22fzx3rzah7pvcyij135yjc3a";
+  };
+
+  nativeBuildInputs = [ autoreconfHook ];
+
+  enableParallalBuilding = true;
```
Comment (gebner) [inline pkgs/os-specific/linux/microcode/iucode-tool.nix]: Typo.

File: pkgs/os-specific/linux/microcode/iucode-tool.nix (hunk 2 above)
Comment (peterhoeg) [reply pkgs/os-specific/linux/microcode/iucode-tool.nix]: Done

File: pkgs/os-specific/linux/microcode/intel.nix (hunk 1 above)
Comment (peterhoeg) [reply pkgs/os-specific/linux/microcode/intel.nix]: Arch Linux includes it (it's a relatively new addition to the microcode distribution) but the problem is that in order to safely apply them, various kernel patches must be in place. I don't know if they are, so I ripped it out again.

File: pkgs/os-specific/linux/microcode/iucode-tool.nix (hunk 3 above)
Comment (peterhoeg) [reply pkgs/os-specific/linux/microcode/iucode-tool.nix]: Thanks

**Review by andir — APPROVED** [review]

**Comment by Moredread** [comment]
I think we should backport this to 18.03, as it contains microcode level fixes for some of the Spectre and Meltdown vulnerabilities.

**Pull request merged**